Introduce separate source file for SSL messaging layer

hanno-arm opened the pull request:
__Summary:__ The goal of this PR is to modify the file structure of the SSL/TLS module in preparation for future changes in both the messaging layer (MPS) and the handshake logic ([D]TLS 1.3):
1. A new file `ssl_msg.c` is introduced which contains the current messaging layer implementation (that is, the record layer and retransmission state machine). This implementation was previously part of `ssl_tls.c,` alongside the implementation of the public SSL API and shared client/server handshake logic functionality.
2. ~Three files `ssl_12_gen.c`, `ssl_12_cli.c` and `ssl_12_srv.c` are introduced which hold handshake reading/writing functions specific to TLS versions <= 1.2 (which is in fact all - for now).~ This is left for a separate PR.

It is the intention of this PR to __solely move code__ but __not modify it__ (except from linkage and private vs [internal] public namespace changes). The hope is that this way it can be integrated quickly, thereby providing a base for future integration of ongoing work on the new messaging layer ~and the TLS 1.3 prototype~ (separate PR). 

Currently, not every function can be clearly attributed to either the messaging or the logic layer, and code adaptation might be needed to improve this. However, this is left for future PRs in order to keep the changes in this PR non-functional.

CC @hannestschofenig

gilles-peskine-arm commented on library/ssl_msg.c:
```diff
@@ -1,7 +1,8 @@
 /*
- *  SSLv3/TLSv1 shared functions
+ *  Generic SSL/TLS messaging layer functions
```
In commit “Move messaging layer functionality to ssl_msg.c”, the commit description says

> the newly created (and thus far empty) ssl_msg.c.

But at this point ssl_msg.c is not empty: it's a copy of ssl_tls.c. Please fix the commit description.

hanno-arm replied:
Fixed.

gilles-peskine-arm commented on include/mbedtls/ssl_internal.h:
```diff
@@ -1064,7 +1064,7 @@ int mbedtls_ssl_decrypt_buf( mbedtls_ssl_context const *ssl,
                              mbedtls_record *rec );
 
```
> Make SSL fn's shared betw'n ssl_tls.c and ssl_msg.c public-internal

Pls dn abbrvt lk ths

“Declare newly shared ssl functions in a header file” would be sufficiently descriptive and is under the (soft) 70 char limit.

A'o g's f'r “Introduce separate source files for TLS 1.2 HS rd/wr functions”

hanno-arm replied:
Was that necessary? Anyway, I'm ok with your suggestion.

gilles-peskine-arm commented on include/mbedtls/ssl_internal.h:
```diff
@@ -1064,7 +1064,7 @@ int mbedtls_ssl_decrypt_buf( mbedtls_ssl_context const *ssl,
                              mbedtls_record *rec );
 
```
How about creating a new header under `library` for these newly shared functions? @mpg @yanesca 

Mbed TLS has no headers under `library`, but Mbed Crypto does and I see no reason why Mbed TLS shouldn't.

mpg replied:
Yes, I'm all for it!

hanno-arm replied:
You're suggesting to move the entire `ssl_internal.h` to `library`, correct? Not just the newly added declarations.

mpg replied:
I'd be happy about moving all `_internal.h` headers to `library`, but I think this should be done in a separate PR. (Also, perhaps after the repo merge will be a bit easier.)

gilles-peskine-arm commented on library/ssl_12_cli.c:
```diff
@@ -1,5 +1,6 @@
- *  SSLv3/TLSv1 client-side functions
+ *  Client-specific handshake reading/writing functions
+ *  for TLS versions <= 1.2.
  *
```
In commit “Populate TLS <= 1.2 handshake source files”:

> This commit moves generic/client/server handshake handling code from ssl_tls.c, ssl_cli.c and ssl_srv.c to the newly created files ssl_12_gen.c, ssl_12_cli.c and ssl_12_srv.c.

Maybe but I can't follow the diff (with or without `--color-moved`). The `_12` files started out as copies of other files. Why is stuff being added to them?

hanno-arm replied:
They started out as entirely unmodified copies for the sake of the git history.

mpg replied:
I understand that the files start as unmodified copies and why, but precisely for that reason I also expect to see only deletions in 22598875438b0534cf3ddf89ff57d094affb15b0 (resp. b769db8c6628ece4feb6fe5b8b1edffa2b178e5b) - before it each piece of code from `ssl_tls.c` exists in two places, so I expect the commit to remove exactly one version of each block, and I'm also surprised to see additions here.

I think it would ease reviewing if the third step in splitting files (after copying to two new files, then renaming one of them to the old name) did nothing else than remove blocks of code from each and the now-duplicated files, and left all the rest (adapting comments in the header, moving common stuff to `ssl_internal.h`, etc.) to later commits.

That way if would be very easy to check programmatically that absolutely no new code is introduced, and that code can only be moved or duplicated - and we can again check programmatically which lines were duplicated (that should be only the first dozen lines in each file: initial comment block with license, plus bunch of `#include` directives).

Right now 22598875438b0534cf3ddf89ff57d094affb15b0 does just a bit too much for me to be conveniently 100% sure it does nothing wrong, while I think that would be easier if all it did was remove blocks of code.

gilles-peskine-arm commented on library/ssl_12_cli.c:
```diff
@@ -1,5 +1,6 @@
 /*
- *  SSLv3/TLSv1 client-side functions
+ *  Client-specific handshake reading/writing functions
+ *  for TLS versions <= 1.2.
```
The `_12` files contain support not only for TLS 1.2 but also for older versions. I find the name misleading: in `ssl_12*`, I expected to find code that is specific to TLS 1.2, not code for TLS 1.1 and 1.0 and SSL 3.0.

hanno-arm replied:
I'm open for suggestions.

mpg replied:
Perhaps `ssl_30-12*.c` or `ssl_30_tls_12*.c`, or `ssl_30_to_tls_12*.c` or `ssl_up_to_tls12*.c`?

And for TLS 1.3 files, perhaps `ssl_tls13*.c`?

Btw, I'm not sure I'm entirely happy with `_gen` either for "generic", as it could also stand for "generated" or "generate". It's obviously a big improvement over the current "tls" which hardly gives a hint, but I'm thinking perhaps `_common` would be better? And while at it, why not change `_cli` and `_srv` to `_client` and `_server`?

hanno-arm replied:
As mentioned below, I prefer to keep the splitting minimal, solely achieving the goal of separating existing protocol logic from messaging logic. I'm fine with splitting further, but I prefer this to be done in a separate PR.

gilles-peskine-arm commented on library/Makefile:
```diff
@@ -75,10 +75,12 @@ OBJS_X509=	certs.o		pkcs11.o	x509.o		\
 		ssl_cli.o	ssl_cookie.o		\
 		ssl_srv.o	ssl_ticket.o		\
-		ssl_tls.o
+		ssl_tls.o	ssl_msg.o
```
Please keep files in lexicographic order

hanno-arm replied:
Fixed.

gilles-peskine-arm commented on library/CMakeLists.txt:
```diff
@@ -83,13 +83,17 @@ set(src_x509
     ssl_srv.c
     ssl_ticket.c
     ssl_tls.c
+    ssl_msg.c
```
Please keep files in lexicographic order

hanno-arm replied:
Fixed.

gilles-peskine-arm commented on library/CMakeLists.txt:
```diff
@@ -174,14 +178,14 @@ endif(USE_STATIC_MBEDTLS_LIBRARY)
 
     add_library(mbedx509 SHARED ${src_x509})
-    set_target_properties(mbedx509 PROPERTIES VERSION 2.20.0 SOVERSION 1)
+    set_target_properties(mbedx509 PROPERTIES VERSION 2.19.1 SOVERSION 1)
```
Blocker: bad merge/rebase: this commit should not touch version numbers (here and below)

hanno-arm replied:
Fixed.

gilles-peskine-arm commented on library/Makefile:
```diff
@@ -37,7 +37,7 @@ endif
 SOEXT_TLS=so.13
 SOEXT_X509=so.1
-SOEXT_CRYPTO=so.4
+SOEXT_CRYPTO=so.3
```
Blocker: bad merge/rebase: this commit should not touch version numbers

hanno-arm replied:
Thanks for spotting.

gilles-peskine-arm commented on library/ssl_12_gen.c:
```diff
@@ -0,0 +1,1126 @@
+#include "mbedtls/oid.h"
+#endif
+
+#if defined(MBEDTLS_SSL_TLS_C)
```
This needs to go before `#include "mbedtls/ssl.h"` (CI is unhappy).

hanno-arm replied:
Fixed.

hanno-arm commented:
@gilles-peskine-arm Thanks for your review, I believe to have addressed your comments.

As for the reviewing, please consider reviewing the cumulative effect of commits 1-3 (from `development` to 2259887) and 5-7 (from bef2d91 to b769db8) through `git diff --color-moved`. Except from a few place where git doesn't detect the moving, this will either make it apparent that code has solely been moved or whether I have slipped somewhere.

gilles-peskine-arm commented:
One DTLS test case (https://jenkins-internal.mbed.com/job/mbedtls-pr-multibranch/job/PR-3010-head/8/execution/node/2593/log/) failed on HEAD but passed on MERGE, passed on other configurations and passes locally. This looks like an intermittent DTLS failure (probably (internal link) https://confluence.arm.com/display/IoTBU/Mbed+TLS+tests#MbedTLStests-DLTStestsconnectionoverbleed), unrelated to this PR.

hanno-arm commented:
@mpg @gilles-peskine-arm @yanesca 

The following diff commands seems useful for review:
* `git diff --histogram --color-moved=dimmed-zebra 36ce88be1edcc9b561bab72b0fd69f6f09b92a1e 22598875438b0534cf3ddf89ff57d094affb15b0` (first bunch of commits splitting `ssl_tls.c` into `ssl_tls.c` and `ssl_msg.c`)
* `git diff --histogram --color-moved=dimmed-zebra bef2d91478ff1103014f632e2c7d80ec9e17b2e2 b769db8c6628ece4feb6fe5b8b1edffa2b178e5b` (second bunch of commits splitting `ssl_{tls|cli|srv}.c` into `ssl_12_{gen|sr|cli}.c` and `ssl_{tls|cli|srv}.c`.

mpg commented on library/ssl_12_cli.c:
```diff
@@ -56,43 +57,163 @@
 
-#if defined(MBEDTLS_KEY_EXCHANGE__SOME__PSK_ENABLED)
-static int ssl_conf_has_static_psk( mbedtls_ssl_config const *conf )
+/* Forward declarations for state-specific handshake functions. */
```
I agree but I would still suggest first leaving the functions in their initial order (ie, refraining from any re-ordering during the copy-rename-trim operation), and then moving them in a separate commit.

hanno-arm replied:
I put functions in the order which I find easiest to read, and I prefer to give the definition of a major function making use of various smaller helpers prior to defining those helpers. This allows to get the big picture before going into the details. At least, that's that's what I personally prefer, but you might think differently of course.

mpg commented:
@gilles-peskine-arm Regarding your plan to ensure that each commit builds, which I like (both the goal and the general plan), it seems to me that steps 5 and 6 need to happen in the same commit. Do you agree? (Btw, steps 3 and 4 could also happen in the same commit without impacting reviewability or buildability I think.)

gilles-peskine-arm commented:
Steps 3 and 4 can happen together, but they're independent so I think they belong in separate commits. Step 5 can be multiple commits. Step 6 can be dissolved into step 5 if it makes more sense.

hanno-arm commented on library/ssl_12_gen.c:
```diff
@@ -61,7345 +54,1073 @@
 
-#if defined(MBEDTLS_SSL_PROTO_DTLS)
+/*
+ * HANDSHAKE STATE: Outgoing `Certificate`
```
The primary goal of this PR is to split existing files in a minimal way so that messaging and logic functionality are separate. I'm in principle fine with further splitting that goes beyond that, but I'd prefer to have this done in a separate PR.

hanno-arm commented on include/mbedtls/ssl_internal.h:
```diff
@@ -1102,4 +1102,8 @@ void mbedtls_ssl_buffering_free( mbedtls_ssl_context *ssl );
 #endif /* MBEDTLS_SSL_PROTO_DTLS */
 
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+void mbedtls_ssl_clear_peer_cert( mbedtls_ssl_session *session );
```
Agreed.

hanno-arm commented:
@gilles-peskine-arm @mpg I have rebased the PR and applied the commit structure that @gilles-peskine-arm recommended. However, I restricted it to the splitting of `ssl_tls.c` into `ssl_tls.c` and `ssl_msg.c`, because I stumbled over points in the further splits about which I wasn't sure on second thought. I'd therefore prefer to prepare the splitting of the remaining `ssl_tls.c` into TLS-<=1.2-specific functions and generic functions (init/free, save/load, config, ...) in a separate PR.

mpg commented:
The CI found a build failure when SSLv3 is enabled (but is otherwise happy):
```
/var/lib/build/library/ssl_msg.c: In function 'mbedtls_ssl_encrypt_buf':
/var/lib/build/library/ssl_msg.c:576:31: error: 'SSL_MAC_MAX_BYTES' undeclared (first use in this function)
             unsigned char mac[SSL_MAC_MAX_BYTES];
                               ^
/var/lib/build/library/ssl_msg.c:576:31: note: each undeclared identifier is reported only once for each function it appears in
/var/lib/build/library/ssl_msg.c:577:13: error: implicit declaration of function 'ssl_mac' [-Werror=implicit-function-declaration]
             ssl_mac( &transform->md_ctx_enc, transform->mac_enc,
             ^
/var/lib/build/library/ssl_msg.c:576:27: error: unused variable 'mac' [-Werror=unused-variable]
             unsigned char mac[SSL_MAC_MAX_BYTES];
                           ^
/var/lib/build/library/ssl_tls.c:1951:13: error: 'ssl_mac' defined but not used [-Werror=unused-function]
 static void ssl_mac( mbedtls_md_context_t *md_ctx,
             ^
```
The fact that `ssl_mac` is seen both as used but not declared _and_ defined but not used probably means it's not defined in the right place. Then it's unclear to me if the issues with `SSL_MAC_MAX_BYTES` and `mac` unused are related or not.

hanno-arm commented:
@gilles-peskine-arm Do you mind addressing the dependencies in subsequent work? After all, one reason why we're reworking the messaging layer is precisely because there are two many interrelations between what _should_ be logically separate.

gilles-peskine-arm commented:
The dependencies are not a blocker, but I'd like to know why you split in a way that has mutual dependencies.

Splits are somewhat disruptive. If we break up the TLS code iteratively, I'd prefer to isolate a few functions that belong together, create a file for them, and then not move them again: subsequent split should address other functions. I'd prefer that to creating a categories, then subdividing each category again and so on.

mpg commented:
@gilles-peskine-arm I suspect completely removing cyclic dependencies may require non-trivial work, in any case work that goes beyond moving functions around to new files. That's because the protocol itself is not quite as cleanly layered as we'd like.

Take for example support for [DTLS clients reconnecting from the same port](https://tools.ietf.org/html/rfc6347#section-4.2.8): at least information needs to be passed both ways between upper (handshake logic) and lower (message handling) layers, since the lower layer needs to ask the upper layer whether this looks like a valid ClientHello with a valid cookie, and to send a HelloVerifyRequest if it's a ClientHello without a valid cookie (and quite obviously the upper layer depends on the lower layer to do that).

Currently this dialog happens by directly calling function of the upper layer from the lower layer, creating a circular dependency. It's probably possible to do it another way (eg the lower layer could return a special error code meaning "can you check this for me, send a HVR if necessary, and call me back when you're done?") but that would be a non-trivial change to a not-entirely trivial feature that's easy to get wrong and hard to fully test (actually I believe the current implementation might have bugs not quite caught by our tests).

So I think I'd also prefer leaving the dependencies for later.

mpg commented on library/ssl_msg.c:
```diff
@@ -1,7 +1,8 @@
 /*
- *  SSLv3/TLSv1 shared functions
+ *  Generic SSL/TLS messaging layer functions
```
I think doing it that way preserves a block of commits in the history where there is absolutely no new line in the code, except as a consequence of copying files. Which makes it a bit easier to make sure that the changes are correct.

mpg replied:
Anyway, I'd rather we avoid rewriting the history at this point unless really necessary.

gilles-peskine-arm replied:
ok

gilles-peskine-arm commented:
Ok for keeping mutual references this time. Please take this into consideration in further splits.

I'll approve if you fix the preprocessor guards in `ssl_internal.h`.

hanno-arm commented on include/mbedtls/ssl_internal.h:
```diff
@@ -1084,4 +1085,21 @@ void mbedtls_ssl_update_out_pointers( mbedtls_ssl_context *ssl,
+
+void mbedtls_ssl_handshake_wrapup_free_hs_transform( mbedtls_ssl_context *ssl );
+
+int mbedtls_ssl_start_renegotiation( mbedtls_ssl_context *ssl );
```
Fixed.

hanno-arm commented on include/mbedtls/ssl_internal.h:
```diff
@@ -1084,4 +1085,21 @@ void mbedtls_ssl_update_out_pointers( mbedtls_ssl_context *ssl,
+
+int mbedtls_ssl_start_renegotiation( mbedtls_ssl_context *ssl );
+
+size_t mbedtls_ssl_get_current_mtu( const mbedtls_ssl_context *ssl );
```
Fixed.

hanno-arm commented on include/mbedtls/ssl_internal.h:
```diff
@@ -1075,6 +1075,7 @@ static inline size_t mbedtls_ssl_ep_len( const mbedtls_ssl_context *ssl )
     return( 0 );
 }
 
+int mbedtls_ssl_resend_hello_request( mbedtls_ssl_context *ssl );
```
Fixed.

hanno-arm commented:
@gilles-peskine-arm Done.

@gilles-peskine-arm @mpg Could you please re-review?
